packages/uploadthing/src/client-future.ts

```diff
@@ -0,0 +1,406 @@
+  progress: number;
+  /**
+   * The key of the file.
+   */
+  key: string;
+  /**
+   * The url of the file.
+   * @example "https://APP_ID.ufs.sh/f/KEY"
+   */
+  url: string;
+  /**
+   * The data returned by the serverside `onUploadComplete` callback.
+   * @example { uploadedBy: "user_123" }
+   */
+  data: TRoute["$types"]["output"];
+  /**
+   * The hash ( <> checksum ) of the file.
+   */
+  hash: string;
+}
+
+export type AnyFile<TFileRoute extends AnyFileRoute> =
+  | PendingFile
+  | UploadingFile
+  | FailedFile<TFileRoute>
+  | UploadedFile<TFileRoute>;
+
+/**
+ * Predicate function to check if a file is pending
+ */
+export function isPendingFile<TRoute extends AnyFileRoute = AnyFileRoute>(
+  file: AnyFile<TRoute>,
+): file is PendingFile {
+  return file.status === "pending";
+}
+
+/**
+ * Predicate function to check if a file is uploading
+ */
+export function isUploadingFile<TRoute extends AnyFileRoute = AnyFileRoute>(
+  file: AnyFile<TRoute>,
+): file is UploadingFile {
+  return file.status === "uploading";
+}
+
+/**
+ * Predicate function to check if a file is failed
+ */
+export function isFailedFile<TRoute extends AnyFileRoute = AnyFileRoute>(
+  file: AnyFile<TRoute>,
+): file is FailedFile<TRoute> {
+  return file.status === "failed";
+}
+
+/**
+ * Predicate function to check if a file is uploaded
+ */
+export function isUploadedFile<TRoute extends AnyFileRoute = AnyFileRoute>(
+  file: AnyFile<TRoute>,
+): file is UploadedFile<TRoute> {
+  return file.status === "uploaded";
+}
+
+/**
+ * Modifies a pending file to an uploading file in place
+ */
+function transitionToUploading(
+  file: PendingFile,
+  rangeStart: number,
+): UploadingFile {
+  const uploadingFile = file as unknown as UploadingFile;
+  uploadingFile.progress = rangeStart;
+  uploadingFile.status = "uploading";
+  return uploadingFile;
+}
+
+/**
+ * Modifies an uploading file to an uploaded file in place
+ */
+function transitionToUploaded<TRoute extends AnyFileRoute>(
+  file: UploadingFile,
+  xhrResult: UploadPutResult,
+): UploadedFile<TRoute> {
+  const uploadedFile = file as unknown as UploadedFile<TRoute>;
+  uploadedFile.status = "uploaded";
+  uploadedFile.data = xhrResult.serverData;
+  uploadedFile.hash = xhrResult.fileHash;
+  uploadedFile.url = xhrResult.ufsUrl;
+  return uploadedFile;
+}
+
+/**
+ * Transitions an uploading file to a failed file in place
+ */
+function transitionToFailed<TRoute extends AnyFileRoute>(
+  file: PendingFile | UploadingFile,
+  reason: UploadThingClientError<TRoute["$types"]["errorShape"]>,
+): FailedFile<TRoute> {
+  const failedFile = file as unknown as FailedFile<TRoute>;
+  failedFile.status = "failed";
+  failedFile.reason = reason;
+  return failedFile;
+}
+
+export interface PresignedReceivedEvent<TRoute extends AnyFileRoute> {
+  type: "presigned-received";
+  /**
+   * The file that caused the event.
+   * @remarks This event is not caused by any one file, so it's the first file in the `files` array.
```

Comment:
is this technical limitation? ideally this could be `null`/`never`?

Reply:
initially thought just having the same API would be nice, but you're right, we can leverage the discriminated union here too 👍